fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OutputPropertyDeriver.java

```diff
@@ -0,0 +1,576 @@
+        }
+    }
+
+    private void transToBucketShuffleJoin(HashDistributionDesc leftLocalDistributionDesc,
+                                          List<Integer> leftShuffleColumns, List<Integer> rightShuffleColumns) {
+        List<Integer> bucketShuffleColumns = Lists.newArrayList();
+        for (int leftScanColumn : leftLocalDistributionDesc.getColumns()) {
+            int index = leftShuffleColumns.indexOf(leftScanColumn);
+            bucketShuffleColumns.add(rightShuffleColumns.get(index));
+        }
+
+        DistributionSpec rightDistributionSpec =
+                DistributionSpec.createHashDistributionSpec(new HashDistributionDesc(bucketShuffleColumns,
+                        HashDistributionDesc.SourceType.BUCKET_JOIN));
+
+        GroupExpression rightChild = childrenBestExprList.get(1);
+        PhysicalPropertySet rightChildOutputProperty = childrenOutputProperties.get(1);
+        // enforce right child BUCKET_JOIN type distribution
+        enforceChildDistribution(rightDistributionSpec, rightChild, rightChildOutputProperty);
+
+        // update group expression require property
+        PhysicalPropertySet newRightChildInputProperty = createPropertySetByDistribution(rightDistributionSpec);
+        requiredChildrenProperties.set(1, newRightChildInputProperty);
+    }
+
+    // enforce child SHUFFLE type distribution
+    private void enforceChildShuffleDistribution(List<Integer> shuffleColumns, GroupExpression child,
+                                                 PhysicalPropertySet childOutputProperty, int childIndex) {
+        DistributionSpec enforceDistributionSpec =
+                DistributionSpec.createHashDistributionSpec(new HashDistributionDesc(shuffleColumns,
+                        HashDistributionDesc.SourceType.SHUFFLE_ENFORCE));
+
+        enforceChildDistribution(enforceDistributionSpec, child, childOutputProperty);
+
+        PhysicalPropertySet newChildInputProperty = createPropertySetByDistribution(enforceDistributionSpec);
+        requiredChildrenProperties.set(childIndex, newChildInputProperty);
+    }
+
+    private void enforceChildDistribution(DistributionSpec distributionSpec, GroupExpression child,
+                                          PhysicalPropertySet childOutputProperty) {
+        double childCosts = child.getCost(childOutputProperty);
+        if (child.getOp() instanceof PhysicalDistributionOperator) {
+            // update child distribution directly
+            PhysicalDistributionOperator rightChildOperator = (PhysicalDistributionOperator) child.getOp();
+            rightChildOperator.setDistributionSpec(distributionSpec);
+            // update child group expression low cost table
+            PhysicalPropertySet newChildOutputProperty = createPropertySetByDistribution(distributionSpec);
+
+            child.setPropertyWithCost(newChildOutputProperty,
+                    child.getInputProperties(childOutputProperty), childCosts);
+            child.getGroup()
```

Comment:
why you could change the best expression of child group in property deriver?

Reply:
enforce new property here, so new property need to set costs in the groupExpression